nslsii/tests/test_xspress3.py

```diff
@@ -0,0 +1,472 @@
+    """
+
+    mcaroi = McaRoi(prefix="MCA1ROI:1:", name="mcaroi01")
+    assert mcaroi.mcaroi_number == 1
+    assert mcaroi.name == "mcaroi01"
+    assert mcaroi.prefix == "MCA1ROI:1:"
+    assert mcaroi.parent is None
+    assert mcaroi.roi_name.pvname == "MCA1ROI:1:Name"
+    assert mcaroi.min_x.pvname == "MCA1ROI:1:MinX"
+    assert mcaroi.size_x.pvname == "MCA1ROI:1:SizeX"
+    assert mcaroi.total_rbv.pvname == "MCA1ROI:1:Total_RBV"
+
+
+def test_sca_init():
+    """
+    Test initialization.
+    """
+
+    sca = Sca(prefix="SCA:", name="sca")
+    assert sca.name == "sca"
+    assert sca.prefix == "SCA:"
+    assert sca.parent is None
+    assert sca.clock_ticks.pvname == "SCA:0:Value_RBV"
+    assert sca.reset_ticks.pvname == "SCA:1:Value_RBV"
+    assert sca.reset_counts.pvname == "SCA:2:Value_RBV"
+    assert sca.all_event.pvname == "SCA:3:Value_RBV"
+    assert sca.all_good.pvname == "SCA:4:Value_RBV"
+    assert sca.window_1.pvname == "SCA:5:Value_RBV"
+    assert sca.window_2.pvname == "SCA:6:Value_RBV"
+    assert sca.pileup.pvname == "SCA:7:Value_RBV"
+    assert sca.event_width.pvname == "SCA:8:Value_RBV"
+    assert sca.dt_factor.pvname == "SCA:9:Value_RBV"
+    assert sca.dt_percent.pvname == "SCA:10:Value_RBV"
+
+
+def test_build_channel_class():
+    """
+    Try to verify all Component attributes are present.
+    """
+    channel_class = build_channel_class(channel_number=2, mcaroi_numbers=(1, 2, 3))
+
+    assert hasattr(channel_class, "channel_number")
+    assert getattr(channel_class, "channel_number") == 2
+
+    assert hasattr(channel_class, "sca")
+    assert hasattr(channel_class, "mca")
+    assert hasattr(channel_class, "mca_sum")
+    assert hasattr(channel_class, "mcarois")
+
+    # there should be 3 MCAROI attributes: mcaroi01, mcaroi02, mcaroi3
+    expected_mcaroi_attr_names = {
+        f"mcaroi{mcaroi_i:02d}" for mcaroi_i in range(1, 3 + 1)
+    }
+
+    # there should be no other MCAROI attributes
+    all_mcaroi_attr_names = {
+        attr_name
+        for attr_name in channel_class.mcarois.__dir__()
+        if attr_name.startswith("mcaroi")
+    }
+
+    assert expected_mcaroi_attr_names == all_mcaroi_attr_names
+
+
+def test_instantiate_channel_class():
+    """
+    Leave verification of Component attributes to the previous test,
+    focus on PV names here.
+    """
+    channel_class = build_channel_class(channel_number=2, mcaroi_numbers=(46, 47, 48))
+    channel_2 = channel_class(prefix="Xsp3:", name="channel_2")
+
+    assert channel_2.sca.clock_ticks.pvname == "Xsp3:C2SCA:0:Value_RBV"
+
+    assert channel_2.mca.array_data.pvname == "Xsp3:MCA2:ArrayData"
+
+    assert channel_2.mca_sum.array_data.pvname == "Xsp3:MCASUM2:ArrayData"
+
+    assert channel_2.mcarois.mcaroi46.total_rbv.pvname == "Xsp3:MCA2ROI:46:Total_RBV"
+    assert channel_2.mcarois.mcaroi47.total_rbv.pvname == "Xsp3:MCA2ROI:47:Total_RBV"
+    assert channel_2.mcarois.mcaroi48.total_rbv.pvname == "Xsp3:MCA2ROI:48:Total_RBV"
```

Comment:
That would be less readable, and I think we should discourage the general use of `getattr`.

Reply:
For tests it's probably OK to spell them explicitly,  but for the main code I wouldn't discourage `getattr`.

Reply:
I'll be more specific and say I think we should discourage use of `getattr` to access the same attributes in more than one place, and encourage defining methods that return or otherwise manipulate the attributes.